lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3928,6 +3926,119 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
+}
+
+void ObjectSubfieldOp::print(OpAsmPrinter &p) {
+  auto input = getInput();
+  auto classType = input.getType();
+  p << ' ' << input << "[";
+  p.printKeywordOrString(classType.getElement(getIndex()).name);
+  p << "]";
+  p.printOptionalAttrDict((*this)->getAttrs(), std::array{StringRef("index")});
+  p << " : " << classType;
+}
+
+ParseResult ObjectSubfieldOp::parse(OpAsmParser &parser,
+                                    OperationState &result) {
+  auto *context = parser.getContext();
+
+  OpAsmParser::UnresolvedOperand input;
+  std::string fieldName;
+  Type inputType;
+  if (parser.parseOperand(input) || parser.parseLSquare() ||
+      parser.parseKeywordOrString(&fieldName) || parser.parseRSquare() ||
+      parser.parseOptionalAttrDict(result.attributes) || parser.parseColon() ||
+      parser.parseType(inputType))
+    return failure();
+
+  if (parser.resolveOperand(input, inputType, result.operands))
+    return failure();
+
+  auto classType = dyn_cast<ClassType>(inputType);
+  if (!classType)
+    return parser.emitError(parser.getNameLoc(),
+                            "input must be !firrtl.class type, got ")
+           << inputType;
+
+  auto index = classType.getElementIndex(fieldName);
+  if (!index)
+    return parser.emitError(parser.getNameLoc(),
+                            "unknown field " + fieldName + " in class type ")
+           << classType;
+  result.addAttribute("index",
+                      IntegerAttr::get(IntegerType::get(context, 32), *index));
+
+  SmallVector<Type> inferredReturnTypes;
+  if (failed(inferReturnTypes(context, result.location, result.operands,
+                              result.attributes.getDictionary(context),
+                              result.getRawProperties(), result.regions,
+                              inferredReturnTypes)))
+    return failure();
+  result.addTypes(inferredReturnTypes);
+
+  return success();
+}
+
+LogicalResult ObjectSubfieldOp::verify() {
+  // Verify that the field exists.
+  auto classType = getInput().getType();
+  auto index = getIndex();
+  if (classType.getNumElements() <= index)
+    return emitOpError() << "field index too large";
+
+  // Verify that the result type of this op agrees with the type of the field.
+  auto type = getType();
+  auto element = classType.getElement(index);
+  if (type != element.type)
+    return emitOpError() << "result type " << type
+                         << " does not match field type " << element.type;
+  return success();
+}
+
+LogicalResult
+ObjectSubfieldOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto classType = getInput().getType();
+  auto className = classType.getNameAttr();
+
+  // Verify that the class exists.
+  auto classOp = dyn_cast_or_null<ClassOp>(
+      symbolTable.lookupSymbolIn(circuitOp, className));
+  if (!classOp)
+    return emitOpError() << "references unknown class " << className;
```

Comment:
maybe add a failure test for this case?

Reply:
To be honest, I'm not sure how to write a failure-test for this case. The class' name is pulled from the type of the input value. Whatever op generated that value will fail to verify first.

Thinking about this more, maybe it isn't the responsibility of this op to verify the type of its input. The op that defines the input will verify that its result is valid. We don't need to recheck it here.

Any dissenting thoughts? I think I can just remove this code.

Reply:
removed